pkg/agent/route/route_windows.go

```diff
@@ -168,6 +176,89 @@ func (c *Client) DeleteRoutes(podCIDR *net.IPNet) error {
 	return nil
 }
 
+func (c *Client) AddServiceRoutes(svcIP net.IP, gwIP net.IP) error {
+	obj, found := c.hostRoutes.Load(svcIP.String())
+	svcIPNet := net.IPNet{IP: svcIP, Mask: net.CIDRMask(32, 32)}
+	metric := metricNormal
+	route := &netroute.Route{
+		LinkIndex:         c.nodeConfig.GatewayConfig.LinkIndex,
+		DestinationSubnet: &svcIPNet,
+		GatewayAddress:    gwIP,
+		RouteMetric:       int(metric),
+	}
+
+	if found {
+		existingRoute := obj.(*netroute.Route)
+		if existingRoute.GatewayAddress.Equal(route.GatewayAddress) && existingRoute.RouteMetric == route.RouteMetric {
+			klog.V(4).Infof("Route for Service IP `%s` already exists via gateway `%s` with metric `%d`",
+				svcIP.String(), gwIP.String(), metric)
+			return nil
+		}
+		// Remove the existing route if gateway or metric is not as expected.
+		if err := c.nr.RemoveNetRoute(existingRoute.LinkIndex, existingRoute.DestinationSubnet, existingRoute.GatewayAddress); err != nil {
+			klog.Errorf("Failed to delete existing route entry with destination `%s` gateway `%s`",
+				existingRoute.DestinationSubnet.String(), existingRoute.GatewayAddress.String())
+			return err
+		}
+	}
+
+	if route.GatewayAddress == nil {
+		return nil
+	}
+
+	// A virtual IP is used to avoid too many ARP responders in OVS flows.
+	// route: SVC IP -> Virtual IP -> antrea-gw0
+	if err := util.NewNetRouteWithMetric(route.LinkIndex, &svcIPNet, GlobalVirtualGWIP, metricNormal); err != nil &&
+		!strings.Contains(err.Error(), "The object already exists") {
+		return err
+	}
+
+	// Route for virtual GW IP.
+	routesVirtualSVCIPNet, err := c.nr.GetNetRoutes(route.LinkIndex, &globalVirtualGWIPNet)
+	if err != nil {
+		return fmt.Errorf("error when getting routes for '%s'", globalVirtualGWIPNet.String())
+	}
+	foundVirtualSVCIPNet := false
+	for _, r := range routesVirtualSVCIPNet {
+		if r.GatewayAddress.Equal(gwIP) {
+			foundVirtualSVCIPNet = true
+			break
+		}
+	}
+	if !foundVirtualSVCIPNet {
+		if err := util.NewNetRouteWithMetric(route.LinkIndex, &globalVirtualGWIPNet, gwIP, metricNormal); err != nil &&
+			!strings.Contains(err.Error(), "The object already exists") {
+			return err
+		}
+	}
+
+	// Remove the Service route on host by kube-proxy.
+	// TODO: When kube-proxy is not used completely, remove the code below.
+	if err := util.RemoveInterfaceAddress("vEthernet (HNS Internal NIC)", svcIP); err != nil {
+		return err
+	}
```

Comment:
I have noticed that if this fail, we do not do any cleanup. Which means that the routes is present on the host, but not in the `hostRoutes` map.

Reply:
Indeed. Added code to clean routes in the `if`.